### DISCUSSION TITLE
Fast Similarity Search for Pattern Detection in Multivariate Time Series

### DISCUSSION THREAD

**Discussion by lorenzo-g-m** [TDAmeritrade/stumpy#346] [Q&A]
I was wondering if the core.mass function (or another equivalent) can be used on multivariate series for fast pattern detection, it isn’t clear to me from the documentation

Many thanks for your help!

**Comment by seanlaw** [comment]
@lorenzo-g-m Welcome to the STUMPY community! Regarding your question, I can interpret it multiple ways so, just so we are on the same page, would you mind telling me what you mean by "fast pattern detection" on "multivariate series"? I understand what [fast pattern search](https://stumpy.readthedocs.io/en/latest/Tutorial_Pattern_Searching.html) is in the case of a 1-D time series.

More specifically, given, say, a multivariate time series with three variables, can you describe what the input(s) would be and what you expect the output to be? And then, how do you plan to use this output?

**Reply by lorenzo-g-m** [reply]
Say you have something like the Activity Recognition system based on Multisensor data fusion (AReM) Data Set in the UCI Machine Learning Repository. Every 0,5 seconds three sensors of activity are recorded, for, say, three months. You know that last Friday at 4pm the person wearing the sensor was doing some “Turbo” abdominals exercise for 3 minutes the doctor has prescribed to her. That 3 minutes constitute your query. You would like to know if and when the same exercise has been performed during last three months, therefore detect the same pattern across the whole time series. Of course the person wearing the sensors has been doing many other activities in the last three months!

**Reply by seanlaw** [reply]
Thank you for the context. So, since each dimension (we restrict our use of “dimensions” to refer only to the number of time series and not to the number of data points) is independent of the other then we would actually call the `core.mass` function three times:

```
import numpy as np

Q = np.random.rand(30).reshape(3, 10)
T = np.random.rand(150).reshape(3, 50)

d, n = T.shape
k = n - m + 1

distance_profile = np.empty((d, k), dtype="float64")

for i in range(d):
    distance_profile[i, :] = core.mass(Q[i], T[i])
```

With this multi-dimensional `distance_profile`, you can then sum up the distances along each column. The column with the smallest sum (or average) is the index location (along `T`) that most closely matches your query, `Q`.

Under the hood, `core.mass` performs some pre-processing (to handle NaN/inf values in `Q` and `T` and also to compute the rolling means and rolling standard deviations for z-normalzation) before computing the `distance_profile`.

**Reply by seanlaw** [reply]
One additional note about `core.mass`. If you have many, many queries (i.e., `Q1`, `Q2`, `Q3`, etc) and you'd like to search for them in `T` then you'll want to avoid continuously pre-processing `T` since its sliding mean/stddev is always the same (unchanging). `core.mass` allows you to provide your own sliding mean/stddev for `T` but you are responsible for pre-processing `T` first:

```
import numpy as np

Q1 = np.random.rand(30).reshape(3, 10)
Q2 = np.random.rand(30).reshape(3, 10)
Q3= np.random.rand(30).reshape(3, 10)
T = np.random.rand(150).reshape(3, 50)

T, M_T, Σ_T = preprocess(T, m)  # This is important especially when `T` contains Nan/inf values!!!

d, n = T.shape
k = n - m + 1

Q1_distance_profile = np.empty((d, k), dtype="float64")
Q2_distance_profile = np.empty((d, k), dtype="float64")
Q3_distance_profile = np.empty((d, k), dtype="float64")

for i in range(d):
    Q1_distance_profile[i, :] = core.mass(Q1[i], T[i], M_T[i], Σ_T[i])
    Q2_distance_profile[i, :] = core.mass(Q2[i], T[i], M_T[i], Σ_T[i])
    Q3_distance_profile[i, :] = core.mass(Q3[i], T[i], M_T[i], Σ_T[i])
```

**Reply by lorenzo-g-m** [reply]
Great news! Many thanks Sean, ciao
